src/Synquid/Parser.hs

```diff
@@ -111,18 +112,36 @@ parseTypeDecl = do
   return $ TypeDecl typeName typeVars typeDef
   where
 
+checkNum :: [Char] -> Bool
+checkNum [] = True
+checkNum (n:ns) = if isDigit n then checkNum ns else False
+
+fixIndex :: RType -> Int
+fixIndex (FunctionT n a r) = 
+  if ("arg" `isPrefixOf` n) 
+    then 
+      (let num = drop 3 n in
+        (if checkNum num 
+          then
+            max (read num::Int) (max (fixIndex a) (fixIndex r))
+          else
+            0))
+    else 
+      max (fixIndex a) (fixIndex r)
+fixIndex _ = 0
+
 fixArgName :: Int -> RType -> RType
-fixArgName _ typ = let (_, res) = fixArgName' 0 typ in res
+fixArgName _ typ = let (_, res) = fixArgName' ((fixIndex typ) + 1) typ in res
   where
     fixArgName' idx (FunctionT x tArg tRes)
-      | "@@arg" `isPrefixOf` x = let
-          (idx', tArg') = fixArgName' (idx + 1) tArg
-          (idx'', tRes') = fixArgName' idx' tRes
-          in (idx'', FunctionT ("arg"++show idx) tArg' tRes')
+      | x == "" = let
+          -- (idx', tArg') = fixArgName' (idx + 1) tArg
+          (idx'', tRes') = fixArgName' (idx + 1) tRes
+          in (idx'', FunctionT ("arg"++show idx) tArg tRes')
       | otherwise = let
-          (idx', tArg') = fixArgName' idx tArg
-          (idx'', tRes') = fixArgName' idx' tRes
-          in (idx'', FunctionT x tArg' tRes')
+          -- (idx', tArg') = fixArgName' idx tArg
```

Comment:
btw, you should delete code you're not using! It's in the git-history, so if we need it again, we can look it up.

Reply:
Removed the old code.